File: vtb-project/matlab/MADmax/MADmax_integer_local_improve.m
function [s,value,nu] = MADmax_integer_local_improve(commonParameters,nu)
% function [s,value,nu] = MADmax_integer_local_improve(commonParameters,nu)
%
% nu is an integer vector
% function checks its feasibility
% if it is feasible, greedy local descent is tried by changing entries by
% +-1
% if it is infeasible only because of the variance constraint, we first try
% to decrease the variance by local descent
% if the solution cannot be further improved, the corresponding variables
% are given back
% value is the maximal achieved objective value
% nu are the solution variables
% s is a boolean indicating whether a feasible solution has been found
% first check feasibility
nu = double(nu);
c = nu'*commonParameters.Gr;
upperwslack = 1 - c;
lowerwslack = c - commonParameters.wmin;
if min(upperwslack,lowerwslack) < 0
    s = false;
    value = [];
    return;
end
old_variance = computeMADvariance(commonParameters,nu);
if old_variance > commonParameters.gammaMAD
    delta_Gr = [commonParameters.Gr; -commonParameters.Gr];
    delta_nu = [eye(length(nu)), -eye(length(nu))];
    while true
        feas_ind = find(([double(commonParameters.numax) - nu; nu] > 0) & (delta_Gr <= upperwslack) & (-delta_Gr <= lowerwslack));
        Mnu = commonParameters.M*nu;
        Mnu_updated = Mnu*ones(1,length(feas_ind)) + commonParameters.M*delta_nu(:,feas_ind);
        variance_updated = mean(abs(Mnu_updated),1);
        [minvar,indminvar] = min(variance_updated);
        if old_variance <= minvar % no improvement in variance
            break;
        else
            old_variance = minvar;
            nu = nu + delta_nu(:,feas_ind(indminvar));
            upperwslack = upperwslack - delta_Gr(feas_ind(indminvar));
            lowerwslack = lowerwslack + delta_Gr(feas_ind(indminvar));
        end
        if old_variance <= commonParameters.gammaMAD % variance constraint satisfied
            break;
        end
    end
end
if old_variance > commonParameters.gammaMAD
    s = false;
    value = [];
    return;
end
s = true;
% start local descent
value = commonParameters.rbar*nu;
% there are n candidate steps, depending on the signs of rbar_i
sign_rbar = sign(commonParameters.rbar);
sign_Gr = (commonParameters.Gr').*sign_rbar;
J = diag(sign_rbar);
while true
    % find variations not violating feasibility
    feas_ind = find((nu'+sign_rbar >= 0) & (nu'+sign_rbar <= double(commonParameters.numax)') & (sign_Gr <= upperwslack) & (-sign_Gr <= lowerwslack));
    for count_feas = length(feas_ind):-1:1
        if computeMADvariance(commonParameters,nu+J(:,feas_ind(count_feas))) > commonParameters.gammaMAD
            feas_ind(count_feas) = [];
        end
    end
    if isempty(feas_ind)
        break; % if we cannot improve the solution we escape the loop
    end
    % choose the index whose change leads to the largest increase in the
    % objective
    [ma,maxind] = max(abs(commonParameters.rbar(feas_ind)));
    value = value + ma;
    ind = feas_ind(maxind);
    nu(ind) = nu(ind) + sign_rbar(ind);
    upperwslack = upperwslack - sign_Gr(ind);
    lowerwslack = lowerwslack + sign_Gr(ind);
end
nu = uint16(nu);

File: src/libraryMIPT/milp/mad_max/MADmax_integer_local_improve.cpp
#include "MADmax_integer_local_improve.hpp"
#include "params.hpp"
#include "utils.hpp"

#include <cmath>

using namespace Eigen;

ReturnValue_localImprove MADmax_integer_local_improve(CommonParams_MADmax* commonParameters, VectorXi32 nu) {
    ReturnValue_localImprove ret;
    VectorXd nu_double = nu.cast<double>();
    double c = nu_double.transpose() * commonParameters->params->Gr;
    double upperwslack = 1 - c;
    double lowerwslack = c - commonParameters->params->w_min;

    if(std::min(upperwslack, lowerwslack) < 0) {
        ret.s = false;
        return ret;
    }

    double old_variance = computeMADvariance(commonParameters, nu_double);
    if(old_variance > commonParameters->gamma_mad) {
        auto Gr_size = commonParameters->params->Gr.size();
        auto nu_size = nu_double.size();

        VectorXd delta_Gr(Gr_size*2);
        delta_Gr << commonParameters->params->Gr, (-1)*commonParameters->params->Gr;

        MatrixXd delta_nu(nu_size, nu_size*2);
        delta_nu << MatrixXd::Identity(nu_size, nu_size), (-1)*MatrixXd::Identity(nu_size, nu_size);
        while(true) {
            VectorXd delta_numax(nu_size*2);
            delta_numax << commonParameters->params->nu_max.cast<double>() - nu_double, nu_double;

            std::vector<int> feas_ind;
            for(auto i = 0; i < nu_size*2; i++) {
                if(delta_numax(i) > 0 && delta_Gr(i) <= upperwslack && -delta_Gr(i) <= lowerwslack)
                    feas_ind.push_back(i);
            }
            VectorXd Mnu = commonParameters->M*nu_double;
            MatrixXd Mnu_update = Mnu*MatrixXd::Ones(1, feas_ind.size()) + commonParameters->M*delta_nu(all, feas_ind);
            RowVectorXd variance_update = Mnu_update.cwiseAbs().colwise().mean();
            int indminvar;
            double minvar = variance_update.minCoeff(&indminvar);
            if(old_variance <= minvar) //no improvement in variance
                break;
            else 
            {
                old_variance = minvar;
                nu_double = nu_double + delta_nu.col(feas_ind[indminvar]);
                upperwslack = upperwslack - delta_Gr(feas_ind[indminvar]);
                lowerwslack = lowerwslack + delta_Gr(feas_ind[indminvar]);
            }
            if(old_variance <= commonParameters->gamma_mad) //variance constraint satusfied
                break;
        }
    }
    if(old_variance > commonParameters->gamma_mad) {
        ret.s = false;
        return ret;
    }

    ret.s = true;
    // start local descent
    ret.value = commonParameters->params->rbar*nu_double;
    //there are n candidate steps, depending on the signs of rbar_i
    Eigen::RowVectorXd sign_rbar = commonParameters->params->rbar.cwiseSign();
    Eigen::RowVectorXd sign_Gr = commonParameters->params->Gr.transpose().cwiseProduct(sign_rbar);
    Eigen::MatrixXd J = sign_rbar.asDiagonal(); //CHECK THIS
    while(true) {
        //find variations not violating feasibility
        std::vector<int> feas_ind;
        for(int i=0; i<sign_rbar.size(); ++i)
        if( ((nu_double.transpose() +sign_rbar)(i) >= 0) &&  
            ((nu_double.transpose() +sign_rbar)(i) <= commonParameters->params->nu_max.transpose().cast<double>()(i)) &&
            (sign_Gr(i) <= upperwslack) && (-sign_Gr(i) <= lowerwslack)) {//OR & <- CHECK
            feas_ind.push_back(i);
        }

        for (auto& count_feas : Utils::matlab_colon<int>(feas_ind.size()-1, 0, -1)) {
            if (computeMADvariance(commonParameters, nu_double + J.col(feas_ind[count_feas])) > commonParameters->gamma_mad) {
                feas_ind.erase(feas_ind.begin() + count_feas); //CHECK!
            }
        }
        if(feas_ind.empty()) {
            break; //if we cannot improve the solution we escape the loop
        }

        //choose the index whose change leads to the largest increase in the objective
        Eigen::RowVectorXd::Index maxind;
        double ma = commonParameters->params->rbar(feas_ind).cwiseAbs().maxCoeff(&maxind);
        ret.value += ma;
        int ind = feas_ind[maxind];
        nu_double(ind)+= sign_rbar(ind);
        upperwslack -= sign_Gr(ind);
        lowerwslack += sign_Gr(ind); 
    }
    ret.nu_local = nu_double.cast<int32_t>();
    return ret;
}